signer/src/stacks/api.rs

```diff
@@ -860,41 +901,41 @@ pub async fn fetch_unknown_ancestors<S, D>(
     stacks: &S,
     db: &D,
     block_id: StacksBlockId,
-) -> Result<Vec<NakamotoBlock>, Error>
+) -> Result<Vec<TenureBlocks>, Error>
 where
     S: StacksInteract,
     D: DbRead + Send + Sync,
 {
-    let mut blocks = vec![stacks.get_block(block_id).await?];
+    let mut blocks = vec![stacks.get_tenure(block_id).await?];
     let pox_info = stacks.get_pox_info().await?;
     let nakamoto_start_height = pox_info
         .nakamoto_start_height()
         .ok_or(Error::MissingNakamotoStartHeight)?;
 
-    while let Some(block) = blocks.last() {
+    while let Some(tenure) = blocks.last() {
         // We won't get anymore Nakamoto blocks before this point, so
         // time to stop.
-        //
-        // TODO: This check is technically incorrect. The nakamoto start height
-        // is the _bitcoin block height_ at which Stacks epoch 3.0 activates.
-        // stacks.get_block() will fail if it's requested a pre-nakamoto block,
-        // but I'm not sure about the stacks.get_tenure() down below.
-        if block.header.chain_length <= nakamoto_start_height {
+        if tenure.anchor_block_height <= nakamoto_start_height {
             tracing::info!(
                 %nakamoto_start_height,
-                last_chain_length = %block.header.chain_length,
+                last_chain_length = %tenure.anchor_block_height,
                 "Stopping, since we have fetched all Nakamoto blocks"
             );
             break;
         }
+        // Tenure blocks are always non-empty, so no need to worr about the
+        // early break.
+        let Some(block) = tenure.blocks.last() else {
+            break;
```

Comment:
Should we log/raise here? Just to be sure we can check if something that shouldn't happen actually happens.

Reply:
I'm going to maintain this invariant in the struct.